sql: apply PCR AOST time stamp for authentication flow

fqazi opened the pull request:
Previously, when using PCR reader catalog its was possible to hit "PCR reader timestamp has moved forward" when attempting to execute workloads. This was because internal executors are exempt from the AOST timestamp. To address this, the authentication code paths are updated to set fixed timestamps, when connecting to a PCR reader catalog. Additionally, additional tests are added for validating opening new connections and preparing queries to confirm this problem is fixed.

Fixes: #135829

Release note: None

fqazi committed:
sql: apply PCR AOST time stamp for authentication flow

Previously, when using PCR reader catalog its was possible to hit "PCR
reader timestamp has moved forward" when attempting to execute
workloads. This was because internal executors are exempt from the AOST
timestamp. To address this, the authentication code paths are updated to
set fixed timestamps, when connecting to a PCR reader catalog.
Additionally, additional tests are added for validating opening new
connections and preparing queries to confirm this problem is fixed.

Fixes: #135829

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135929)
<!-- Reviewable:end -->

msbutler commented on pkg/sql/catalog/replication/reader_catalog_test.go:
```diff
@@ -322,6 +328,16 @@ func TestReaderCatalogTSAdvance(t *testing.T) {
 			require.NotEqualValues(t, srcRes, destRes)
 		}
+
+		// Sanity: Execute the same query as prepared statement inside the reader
```
to confirm, were these unit tests with prepared stmts failing before your patch to the source code?

fqazi replied:
I didn't see any issues on the prepared statements, but opening new connections could fail intermittently.
<!-- Reviewable comment -OCJVU7N5_zkt0EhDOJb:b3grkjp -->
<!-- Sent from Reviewable.io -->

msbutler commented on pkg/sql/catalog/replication/reader_catalog_test.go:
```diff
@@ -432,7 +473,7 @@ func TestReaderCatalogTSAdvance(t *testing.T) {
 func TestReaderCatalogTSAdvanceWithLongTxn(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.UnderDuress(t)
+	//	skip.UnderDuress(t)
```
nit: intended?

msbutler reviewed (commented):
thanks for the patch!! I'll try running my roachtest on top of this. I'll let someone from foundations approve this. 

One other thought: I wonder if there's a more general way to find other internal executor read queries that require a fixed timestamp. Or maybe this points to a larger design problem: perhaps no system tables should be externalized. cc @dt

fqazi committed:
catalog/lease: mismatchedExternalDataRowTimestamp retryable

Previously, when mixed ExternalDataRow time stamps were detected between
descriptors during a txn, an error would be returned. While user
operations should never see this error because of an AOST set on them
(to guarantee a consistent view of descriptors), internal operations
could observe this error. To address this, this patch marks the
mismatchedExternalDataRowTimestamp error as retryable. This will cause
internal txns to retry if the error gets hit.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler)



<!-- Sent from Reviewable.io -->

msbutler commented on pkg/sql/user.go:
```diff
@@ -288,92 +291,98 @@ func retrieveAuthInfo(
-	aInfo.CanLoginDBConsoleRoleOpt = true
-	var ok bool
+		roleOptsIt, err := txn.QueryIteratorEx(
+			ctx, "get-login-dependencies", txn.KV(), /* txn */
```
you're fixing this in this PR, but it seems a bit odd that we previously grabbed the login dependencies in a separate txn than the hashed password.

fqazi replied:
Yeah,  I think it was just a subtle miss. I don't think it was because the txn complexity would be too high here, since their both scanning for one row.
<!-- Reviewable comment -OCYpvAa92Ksi29Bobw_:b-j3cumm -->
<!-- Sent from Reviewable.io -->

fqazi replied:
Done.

Also added:

// We are going to use a single txn for resolving the user information,  
// and role\_options for the user.
<!-- Reviewable comment -OCfS9-aCPjnAVXJlxEM:b-bisfbo -->
<!-- Sent from Reviewable.io -->

msbutler commented on pkg/sql/catalog/replication/reader_catalog_test.go:
```diff
@@ -409,6 +410,9 @@ func TestReaderCatalogTSAdvance(t *testing.T) {
 			if !useAOST {
+				// Toggle the block on each iteration, so there is some risk
+				// of not all descriptor updates being updated.
+				descriptorRefreshHookEnabled.Swap(true)
```
I'm finding it hard to reason through all the components of this test, perhaps because i'm unfamiliar with the LeaseManager hooks you're using. Perhaps in a follow up PR after we fix the bug, we could break this test up into a few smaller tests that are easier to understand.

fqazi replied:
Fair point, and I think a bunch of code can be shared here too. I opened: #136114 for tracking
<!-- Reviewable comment -OCYqfGiA5HJ-v77CniN:b-eqox7r -->
<!-- Sent from Reviewable.io -->

msbutler reviewed (approved):
Thanks again for doing this! Also, are you ok backporting this to 24.3?

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler)



<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-:-OCd7_n676eotCcs0IML:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @msbutler)

___
*[`pkg/sql/user.go` line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6-Sb89JbyZkXYUMQ:-OCd6-Sb89JbyZkXYUMR:b-ktg4kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L290)):*
> ```Go
> ) (aInfo sessioninit.AuthInfo, retErr error) {
> 	// Use fully qualified table name to avoid looking up "".system.users.
> 	// We use a nil txn as login is not tied to any transaction state, and
> ```

can you update this comment to explain that we use a DescsTxn and have it describe why we need `MaybeSetReplicationSafeTS`?
___
*[`pkg/sql/user.go` line 303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd764a-TLKlWpfFftk:-OCd764bC8fvOxPfjHkC:b-3jhmdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L303)):*
> ```Go
> 			getHashedPassword, user)
> 		if err != nil {
> 			return err
> ```

nit: why remove the `errors.Wrap` here?
___
*[`pkg/sql/user.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6dvM8XWlkNe1rDe4:-OCd6dvM8XWlkNe1rDe5:b-qd1vi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L349)):*
> ```Go
> 		var ok bool
> 
> 		for ok, err = roleOptsIt.Next(ctx); ok; ok, err = roleOptsIt.Next(ctx) {
> ```

could you introduce a `var loopErr error` variable? it makes the error handling a bit easier to read
___
*[`pkg/sql/user.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6kZHDCIC93_sO11i:-OCd6kZHDCIC93_sO11j:bxtvnub) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L384)):*
> ```Go
> 			}
> 		}
> 		return err
> ```

super nit: change to 
```
if loopErr != nil {
  return loopErr
}
return nil
```


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @rafiss)

___
*[`pkg/sql/user.go` line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6-Sb89JbyZkXYUMQ:-OCfRC7uDKUraYhBES2l:bvhzo6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L290)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you update this comment to explain that we use a DescsTxn and have it describe why we need `MaybeSetReplicationSafeTS`?
</blockquote></details>

Done. 

Added the comment:

// When running on a PCR reader catalog we need to ensure all descriptors  
// have matching timestamps for external row data. Certain system tables  
// like users and role\_options are replicated, which can cause mixed  
// external row data timestamps, which can lead to a retryable error.  
// To avoid this we will set a replication safe AOST timestamp when running  
// on a reader catalog.
___
*[`pkg/sql/user.go` line 303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd764a-TLKlWpfFftk:-OCfPokoB4GvfZEJ061b:bqstdq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L303)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: why remove the `errors.Wrap` here?
</blockquote></details>

Done.

That was accidental, good catch.
___
*[`pkg/sql/user.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6dvM8XWlkNe1rDe4:-OCfQNHT63D2m-qJWESV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L349)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you introduce a `var loopErr error` variable? it makes the error handling a bit easier to read
</blockquote></details>

Done.
___
*[`pkg/sql/user.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCd6kZHDCIC93_sO11i:-OCfQSEJ6BlsvcF03zlt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ebda72436fb188f432a86d97ccc46bf82777023/pkg/sql/user.go#L384)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: change to 
```
if loopErr != nil {
  return loopErr
}
return nil
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @msbutler)

___
*[`pkg/sql/user.go` line 290 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/135929#-OCjVKQR-h6OKf6yqKl7:-OCjVKQR-h6OKf6yqKl8:b-l51k8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6659ceca3803cc3938b1bbc9bc0a9d37d0202a1c/pkg/sql/user.go#L290)):*
> ```Go
> ) (aInfo sessioninit.AuthInfo, retErr error) {
> 	// Use fully qualified table name to avoid looking up "".system.users.
> 	// We use a nil txn as login is not tied to any transaction state, and
> ```

super nit: remove the words about using a `nil txn`


<!-- Sent from Reviewable.io -->

fqazi commented:
@msbutler @rafiss TFTR!

bors r+
